File: SDKSampleApp/app/src/java/java/com/fourthline/sdksample/SelfieActivity.java
/*
 * Copyright © 2020 Safened - Fourthline B.V. All rights reserved.
 */
package com.fourthline.sdksample;

import static android.widget.Toast.LENGTH_LONG;
import static androidx.lifecycle.LifecycleOwnerKt.getLifecycleScope;
import static com.fourthline.sdksample.SelfieResultActivity.KEY_SELFIE_VIDEO_URI;
import static com.fourthline.sdksample.Utils.asString;
import static com.fourthline.vision.selfie.LivenessCheckType.HEAD_TURN;
import static kotlin.collections.CollectionsKt.joinToString;

import android.content.Context;
import android.content.Intent;
import android.graphics.Rect;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import com.fourthline.vision.RecordingType;
import com.fourthline.vision.selfie.SelfieScannerActivity;
import com.fourthline.vision.selfie.SelfieScannerConfig;
import com.fourthline.vision.selfie.SelfieScannerError;
import com.fourthline.vision.selfie.SelfieScannerResult;
import com.fourthline.vision.selfie.SelfieScannerStep;
import com.fourthline.vision.selfie.SelfieScannerWarning;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.net.URI;
import java.util.List;

public class SelfieActivity extends SelfieScannerActivity {

    private static final String TAG = SelfieActivity.class.getSimpleName();

    private View overlayView;
    private View selfieMask;
    private PunchholeView punchhole;
    private TextView stepLabel;
    private TextView warningsLabel;

    @NotNull
    @Override
    public SelfieScannerConfig getConfig() {
        return new SelfieScannerConfig(
                false, // debugModeEnabled
                RecordingType.VIDEO_ONLY,
                HEAD_TURN
        );
    }

    @Nullable
    @Override
    public View getOverlayView() {
        overlayView = LayoutInflater.from(this).inflate(
                R.layout.overlay_selfie,
                findViewById(android.R.id.content),
                false);

        return overlayView;
    }

    @NotNull
    @Override
    public Rect getFaceDetectionArea() {
        return new Rect(selfieMask.getLeft(),
                selfieMask.getTop(),
                selfieMask.getRight(),
                selfieMask.getBottom()
        );
    }


    @Override
    protected void onCreate(@Nullable Bundle bundle) {
        super.onCreate(bundle);

        selfieMask = overlayView.findViewById(R.id.selfieMask);
        punchhole = overlayView.findViewById(R.id.punchhole);
        stepLabel = overlayView.findViewById(R.id.stepLabel);
        warningsLabel = overlayView.findViewById(R.id.warningsLabel);
        overlayView.findViewById(R.id.buttonClose).setOnClickListener((view) -> {
            Utils.vibrate(this);
            finish();
        });

        Utils.onLayoutMeasuredOnce(getWindow().findViewById(android.R.id.content), (view) -> {
            punchhole.setPunchholeRect(getFaceDetectionArea());
            punchhole.postInvalidate();
            return null;
        });
    }

    /**
     * You may enter {@link SelfieScannerStep#MANUAL_SELFIE} step when {@link SelfieScannerConfig#getIncludeManualSelfiePolicy()}
     * is set to TRUE and selfie step timeouts.
     * In this case present snapshot button to user and call {@link com.fourthline.vision.selfie.SelfieScanner#takeSnapshot} on click.
     */
    @Override
    public void onStepUpdate(@NotNull SelfieScannerStep step) {
        Utils.vibrate(this);
        Log.d(TAG, "Step update: " + step);

        runOnUiThread(() -> stepLabel.setText(asString(step)));
    }

    @Override
    public void onWarnings(@NotNull List<? extends SelfieScannerWarning> warnings) {
        String prettyWarnings = joinToString(warnings, ", ", "[ ", " ]", -1, "...", null);
        Log.d(TAG, "Current warnings: " + prettyWarnings);

        runOnUiThread(() -> warningsLabel.setText(asString(warnings.get(0))));

        Utils.scheduleCleanup(getLifecycleScope(this), warningsLabel);
    }

    @Override
    public void onFail(@NotNull SelfieScannerError error) {
        Utils.vibrate(this);
        Log.d(TAG, "Scanner failed with error: " + error);

        runOnUiThread(() -> Toast.makeText(this, asString(error), LENGTH_LONG).show());

        finish();
    }

    @Override
    public void onSuccess(@NotNull SelfieScannerResult result) {
        Utils.vibrate(this);
        Log.d(TAG, "Selfie scan succeed");

        Bundle data = new Bundle();
        URI videoUri = result.getVideoUrl();
        data.putString(KEY_SELFIE_VIDEO_URI, videoUri != null ? videoUri.getPath() : "n/a");

        SelfieResultActivity.start(this, data, result.getImage().getCropped());

        finish();
    }

    public static void start(Context context) {
        Intent intent = new Intent(context, SelfieActivity.class);
        context.startActivity(intent);
    }
}

File: SDKSampleApp/app/src/kotlin/java/com/fourthline/sdksample/activities/SelfieActivity.kt
/**
 * Copyright © 2020 Safened - Fourthline B.V. All rights reserved.
 */
package com.fourthline.sdksample.activities

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Log
import com.fourthline.sdksample.*
import com.fourthline.sdksample.component.SelfieScannerOverlayView
import com.fourthline.sdksample.helper.selfiescanner.*
import com.fourthline.vision.selfie.*
class SelfieActivity : SelfieScannerActivity() {

    private val TAG: String = SelfieActivity::class.java.simpleName

    /**
     * Creates and keeps an instance of custom view SelfieScannerOverlayView, which is passed to
     * the required method [SelfieScannerActivity.getOverlayView].
     *
     * The Fourthline SDK provides functionality for efficient scanning, while offering you the
     * flexibility to customize the experience to fit your product and identity.
     *
     * In this example application, the delegation pattern is used to give some more flexibility
     * to this view. This prevents some code-duplication as this view is used for both the
     * Activity and Fragment.
     */
    private val scannerOverlayView by lazy {
        SelfieScannerOverlayView(this).apply {
            configureButtons(onCloseScanner = {
                vibrate()
                finish()
            })
        }
    }

    /**
     * Overlay view for the Camera, containing buttons as well as guidance for the end-user.
     */
    override fun getOverlayView() = scannerOverlayView

    /**
     * Coordinates within the Camera Viewport used for Face Detection.
     */
    override fun getFaceDetectionArea() = scannerOverlayView.getFaceDetectionArea()

    /**
     * Configuration for the [SelfieScannerActivity].
     */
    override fun getConfig() = SelfieScannerConfig()

    /**
     * Configure the Activity.
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Clear previous Selfie result data on KycInfo
        KycResultContainer.selfieResult = null
    }

    /**
     * You may enter [SelfieScannerStep.MANUAL_SELFIE] step when [SelfieScannerConfig.includeManualSelfiePolicy]
     * is set to TRUE and selfie step timeouts.
     * In this case present snapshot button to user and call [SelfieScannerActivity.takeSnapshot] on click.
     */
    override fun onStepUpdate(step: SelfieScannerStep) = runOnUiThread {
        vibrate()
        Log.d(TAG, "Step update: $step")

        scannerOverlayView.onStepUpdate(step)
    }

    /**
     * Handle warnings during scanning, as returned by [SelfieScannerActivity].
     *
     * Informing the end-user will help them overcome any challenges, increasing the likeliness of
     * successful selfie scanning.
     */
    override fun onWarnings(warnings: List<SelfieScannerWarning>) = runOnUiThread {
        val prettyWarnings = warnings.joinToString(", ", "[", "]")
        Log.w(TAG, "Current warnings: $prettyWarnings")

        scannerOverlayView.onWarnings(warnings)
    }

    /**
     * Handle errors returned by [SelfieScannerActivity].
     */
    override fun onFail(error: SelfieScannerError) = runOnUiThread {
        vibrate()
        Log.e(TAG, "Scanner failed with error: $error")

        showDialog(
            title = getString(R.string.common_failure),
            message = getString(error.getErrorLabel()),
            positiveButtonText = getString(R.string.common_try_again),
            positiveButtonHandler = { recreate() },
            negativeButtonText = getString(R.string.common_dismiss),
            negativeButtonHandler = { finish() }
        )
    }

    /**
     * Handle successful [SelfieScannerResult] returned by [SelfieScannerActivity].
     */
    override fun onSuccess(result: SelfieScannerResult) {
        vibrate()
        Log.e(TAG, "Selfie scan succeed")

        // Save results for later reference when submitting KycInfo for validation
        KycResultContainer.selfieResult = result

        SelfieResultActivity.start(this)

        finish()
    }

    /**
     * Convenience launcher for Activity.
     */
    companion object {
        fun start(context: Context) = with(context) {
            Intent(this, SelfieActivity::class.java).let(::startActivity)
        }
    }
}